content/en/blog/_posts/2020-08-04-leaving-the-swarm-road-to-kubernetes.md

```diff
@@ -0,0 +1,172 @@
+
+We also recognized that the migration from 1.7 to 2.x required a new approach to configuration and since we were already migrating from Swarm to Kubernetes we should just go ahead and go with the latest version now.
+
+With the upgrade, we had the option of leveraging the existing Kubernetes Ingress pattern along with annotations or leverage CRDs -- for us, it made sense to use the CRD option as it reduced the clutter of adding and managing a slew of conditional annotations in our manifests.
+
+#### Migrating Infrastructure
+
+Prior to Kubernetes, we used single instances of [Redis](https://redis.io/) to handle caching and locking concerns, and [RabbitMQ](https://www.rabbitmq.com/) for async task queuing. At some point in the next year or two, our workload will require scaling those technologies out just to handle our projected transactional load. We are ramping up our analytics platform as well and have begun implementing [Kafka](https://kafka.apache.org/) and [Redshift](https://aws.amazon.com/redshift/) for that project.
+
+Our production workloads completely rely on managed services for our transactional datastores (RDS, Redshift, Kafka), but our non-production environments leverage the open-source equivalents. Moving to Kubernetes means having the option of leveraging operators for those technologies when we will eventually have to handle scaling out RabbitMQ and Redis in production, as well as giving our non-production environments operational equivalents for our transactional components. We had no trouble implementing the bitnami charts for RabbitMQ and Redis while using the official [confluent helm chart](https://github.com/confluentinc/cp-helm-charts) for Kafka.
+
+We haven’t begun scaling out those components yet, but the pieces are in place for us to begin experimenting with them locally first, then eventually moving to high availability in production. Having both the commercial and community support for this capability didn’t exist without first moving to Kubernetes.
+
+#### Ready ... Set ... Oops ...
+
+We had met our benchmarks for the development environment, had managed to stand up our platform in staging on EKS, and had successfully migrated our critical infrastructure components and worked out some kinks we’d found.
+
+One notable issue we’d discovered was when we capture user geo-location data the service began to fail when processing the transaction. Modern load balancers generate an `X-Forwarded-For` header that allows the destination service to authenticate the transaction by validating the subnets of the proxies in the request chain, but it’s also the source of truth for our reporting system. At first, we just simply assigned the cause of the missing header to the oddities of Metallb and the local environment, but we quickly discovered the issue was also present in the staging environment in EKS, in what was eventually going to be our production environment.
```

Comment:
Possibly:
```suggestion
One notable issue we’d discovered was when we capture user geo-location data the service began to fail when processing the transaction. Load balancers commonly generate an `X-Forwarded-For` header that allows the destination service to authenticate the transaction by validating the subnets of the proxies in the request chain, but it’s also the source of truth for our reporting system. At first, we just simply assigned the cause of the missing header to the oddities of Metallb and the local environment, but we quickly discovered the issue was also present in the staging environment in EKS, in what was eventually going to be our production environment.
```

because `X-Forwarded-For:` isn't particularly modern - it's from the 1990s.